README.md
# PlAcER: Plastic Accumulation Estimate using R


> ## Overview

A computational toolbox for performing simulations on the prevalence of debris in bird's nests, based on bootstrap replicates. The package allows for calculating bootstrapped 95% confidence intervals (CI) for the estimated prevalence of debris. CIs are taken as the 2.5 and 97.5 percentiles of the distributions resulting from each sample size (Efron & Tibshirani, 1993). Combined with a Bayesian approach for detecting change points in a curve with values of prevalence as functions of sample size, i.e. number of nests, the resampling simulations can be also used to define appropriate sample sizes to detect prevalence of plastics. The method has wide application, and can also be applied to estimate confidence intervals and define sample sizes for the prevalence of plastics ingested by any other organisms.

> ## Features

Efficient for performing simulation on how accurate is the estimated prevalence of plastic debris accumulated in any kind of organism or environment by using bootstrap replicates. This method works regardless of any data distribution assumptions.

Allows for assessing how many samples should be considered for obtaining an appropriate accuracy level for the estimated prevalence of debris desired in a given application.

> ## Installation

To install the package, type in the R console:

```r
install.packages('placer')
```

> ## Usage

Load the pacakge and the data containing observations of presence of plastic in nests of Caspian terns. This data comprises 529 nests of Caspian terns observed in Senegal, where “yes” = debris present and “no” = debris absent. Data obtained from Tavares et al. (2019).

```r
library("placer")
data(ctern)
```

Calculate the bootstrap prevalence probability of plastic based on a specific
number of samples. Then from the resampled data the 95% confidence intervals (CI)
are computed. To quantify the prevalence probability and the CI simply provide
the presence/absense data, the maximum number of sample sizes to test (default 300),
and the number of bootstrap simulations (default 1000) as:

```r
ternsci<-placer::plastic.ci(cterns$debris_presence, 300, 1000)
```
Where the user must first specify the presence/absence data (cterns$debris_presence), followed by he maximum number of nests considered (300), and last the number of replicates (1000). Consider that by increasing the maximum sample size the computational time substantially increases.

Plot the plastic preevalence probability and the confidence interval 
as function of sample size: 

```r
placer::prevalence_plot(ternsci$prevprob,
ternsci$cidtf$N,
ternsci$cidtf$lower_ci,
ternsci$cidtf$upper_ci)
```

Where the user must firts specify the calculated prevalence probability as obtained from plastic.ci function, then plot the correspoing sample size, confidence interval values. 

> ## Where to get help

For questions or issues about placer usage and more general questions on bootstrap simulations and applications to frequency data please use github repository (https://github.com/systemsecologygroup/PlAcE_R).

> ## References

Efron, B., & Tibshirani, R. (1993). An introduction to the Bootstrap. Boca Raton: Chapman & Hall.

Tavares, D. C., Moura, J. F., & Merico, A. (2019). Anthropogenic debris accumulated in nests of seabirds in an uninhabited island in West Africa. Biological Conservation, 236, 586–592.

Tavares, D.C., Moura, J.F., Acevedo-Trejos, E., Crawford, M., Makhado, A., Lavers, J.L., Witteveen, M., Ryan, P., Merico, A. Submitted.


NEWS.md
# News: Plastic Accumulation Estimate using R (PlAcER)

## PlAcER version 0.1.x
Initial release of the package 'placer'.

### Bug fixes
* 0.1.1: Corrected warnings and notes for building the package.
* 0.1.2: Corrected warnings and notes for building the package.
* 0.1.3: Fixes to complete CRAN submission.


placer/R/placer.R
#' placer: A package to estimate the accumulation of plastic debris in bird's nests
#'
#' The package placer consist of two main funtions to estimate the accumulation of plastic
#' in bird's nest as a function of sample size, and a rutine to plot the plastic prevalence
#' probability, and thier confidence intervals as a function of sample size.
#'
#' @section Main functions:
#' The function \code{\link{plastic.prev.prob}} calculates the plastic
#' prevalence probability for a given sample size based on presence and abscence data
#' The function \code{\link{plastic.ci}} estimates the 95\% confidence intervals for a given
#' prevalence probability of plastic debris. In addition, the package placer includes a plotting
#' rutine \code{\link{prevalence_plot}} to show the estimated plastic prevalence probability and
#' thier 95\% confidence intervals as a function of sample size.
#'
#' @docType package
#' @name placer
NULL


placer/R/plastic_prev_plots.R
#------------------------------------------------------------------------------
# Plotting routine for the plastic prevalence probability
#
# Developed by: Davi Castro Tavares
# Date: May 2018
# Modified by: Esteban Acevedo-Trejos
# Date: July 2019
#------------------------------------------------------------------------------

#' Plastic prevalence probability plot
#'
#' Plot to show the plastic prevalence probability in seabird's nests as a function of
#' different sample sizes and their corresponding confidence intervals.
#'
#' @param prev_prob_mat numeric matrix, containing plastic prevalence probability
#'   with dimensions (samples_size, bootstrap_replicates).
#' @param sample_sizes numeric vector, containing sequence of sample size used to
#'   estimate the confidence intervals \code{plastic.ci}.
#' @param lower_ci numeric vector, containing values for lower confidence interval
#'   and with the same length as \code{sample_sizes}.
#' @param upper_ci numeric vector, containing values for upper confidence interval
#'   and with the same length as \code{sample_sizes}.
#' @param xlab string, label of x axis.
#' @param ylab string, label of y axis.
#' @param colobs color of observations.
#' @param colci color of confidence intervals.
#' @seealso \code{\link{plastic.ci}}, \code{\link{plastic.prev.prob}}
#' @examples
#' binomtest <- plastic.ci(rbinom(1000,1,0.5), 30, 100)
#' prevalence_plot(binomtest$prevprob,
#'                 binomtest$cidtf$N,
#'                 binomtest$cidtf$lower_ci,
#'                 binomtest$cidtf$upper_ci)
#' @importFrom graphics matplot par points
#' @export
prevalence_plot <- function(prev_prob_mat, sample_sizes, lower_ci, upper_ci,
                            xlab = "Sample size",
                            ylab = "Plastic prevalence probability",
                            colobs = "grey", colci = "#64B5F6"){
  oldpar <- par(no.readonly = T)
  #defining plot (quadrant) parameters.
  par(mar = c(5, 4, 4, 4) + 0.3)
  #defining a sequence from 1 to 300 to the horizontal axis, and calling the simulated prevalence values (object called output).
  matplot(sample_sizes, prev_prob_mat, type = "l", lty = 1, col = colobs,
          yaxp = c(0, 1, 5), cex.lab = 0.9, tck = -0.03, cex.axis = 0.9,
          las = 1, xlab = xlab,
          ylab = ylab)
  #adding lower CIs to the plot.
  points(sample_sizes, lower_ci, type = "p", cex = 0.2, ylim = c(0,1), col = colci)
  #adding upper CIs to the plot.
  points(sample_sizes, upper_ci, type = "p", cex = 0.2, ylim = c(0,1), col = colci)
  par(oldpar)
}


placer/R/plastic_estimation.R
#------------------------------------------------------------------------------
# Main functions to estimate plastic prevalence probability in seabird's nests
# as a function of different sample sizes and thier corresponding confidence
# intervals.
#
# Developed by: Davi Castro Tavares
# Date: May 2018
# Modified by: Esteban Acevedo-Trejos
# Date: July 2019
#------------------------------------------------------------------------------


#' Confidence intervals of plastic prevalence probability
#'
#' Bootstrap simulations to estimate 95\% bootstrapped CIs for the
#' prevalence of debris obtained with different sample sizes.
#'
#' @param plastic_abs_pres numeric vector, containing a binary values with 0
#'   or no for absence of plastic, and 1 or yes for presence of plastic.
#' @param max_sample_size integer, specifying the maximum number of
#'   samples to use for estimating the prevalence of plastic debris.
#'   By default 300 samples. Increasing sample sizes substantially increases
#'   compuational time.
#' @param bs_rep integer, specifying the number of bootstrap replications.
#'   By default 1000 replications.
#' @param lower_ci numeric, specifying lower confidence interval.
#'   By default 2.5\%, based on Efron and Tibshirani (1993)
#' @param upper_ci numeric, specifying upper confidence interval.
#'   By default 97.5\% default, based on Efron and Tibshirani (1993).
#' @return A list (cidtf) with a data frame with sample sizes, mean CI,
#' lower CI, upper CI, and a matrix (prevprob) with prevalence probability
#' of plastic debris for all sample sizes and their estimated prevalence of debris.
#' @references Efron, B., & Tibshirani, R. (1993). An introduction to the Bootstrap.
#'  Boca Raton: Chapman & Hall.
#' @seealso \code{\link{plastic.prev.prob}}, \code{\link{prevalence_plot}}
#' @note The confidence intervals are calculated in a sequence of varying
#' sample sizes, i.e. 1,2,3...,n and the function can be also used for defining
#' sample sizes that would provide 95\% CIs with the desired accuracy.
#' @examples
#' plastic.ci(rbinom(1000,1,0.5), 30, 100)
#' @export
plastic.ci <- function(plastic_abs_pres, max_sample_size = 300, bs_rep = 1000,
                       lower_ci = 0.025, upper_ci = 0.975){

  outmat <- matrix(nrow = max_sample_size, ncol = bs_rep)
  for(ss in 1:max_sample_size){
    outmat[ss,] <- replicate(bs_rep, plastic.prev.prob(plastic_abs_pres, ss))
  }
  ci <- apply(outmat, 1, function(x) quantile(x, probs = c(lower_ci, upper_ci)))
  mean_ci <- apply(outmat, 1, function(x) mean(x))
  cidtf <- data.frame(N = 1:max_sample_size, mean_ci = mean_ci,
                      lower_ci = ci[1,], upper_ci = ci[2,])

  return(list(cidtf=cidtf, prevprob=outmat))
}


#' Plastic prevalence probability
#'
#' \code{plastic.prev.prob} estimates the prevalence probability of plastic
#' from a randomly selected sample of absence/presence observations of plastic
#' debris.
#'
#' @param plastic_abs_pres numeric vector, containing a binary values with 0
#'   or 'no' for absence of plastic, and 1 or 'yes' for presence of plastic.
#' @param num_sample integer value, specifying the number of samples to
#'   randomly draw from the observations.
#' @return Prevalence probability of plastic debris in a given sample size.
#' @seealso \code{\link{plastic.ci}}, \code{\link{prevalence_plot}}
#' @examples
#' plastic.prev.prob(rbinom(1000,1,0.5), 1)
#' plastic.prev.prob(rbinom(1000,1,0.5), 10)
#' @importFrom stats quantile
#' @export
plastic.prev.prob <- function(plastic_abs_pres, num_sample){
  if(length(unique(plastic_abs_pres)) != 2){
    # check if a given variable is binary and in the adequate binary format
    stop('input variable plastic_abs_pre is not a binary varible')
  } else if(all(sort(unique(plastic_abs_pres)) != c(0,1)) &
            all(sort(unique(plastic_abs_pres)) != c('no','yes'))){
    stop('values of bynary variable must be 0 or no (for absence)
         and 1 or yes (for presence)' )
  } else {
    # Draw a random sample from the observations of plastic debris and
    # calculates the prevalence probability of plastic in this sample
    rand_sample <- dplyr::sample_n(data.frame(plastic_abs_pres),
                                   num_sample, replace = TRUE)
    rand_sample_pres <- subset(rand_sample,
                               rand_sample == 1 | rand_sample == 'yes')
    plastic_prev <- nrow(rand_sample_pres) / nrow(rand_sample)
    return(plastic_prev)
  }
}

